testsuite/tests/templates/basic/test.ml

```diff
@@ -530,58 +754,286 @@
+           check-ocamlopt.byte-output;
          }{
-           flags = "-as-parameter";
-           module = "category_b.mli";
+           module = "";
+           flags = "-instantiate";
+           program = "import-Int_list_element.cmx";
+           all_modules = "import.cmx int_list_element.cmx";
+           compiler_output = "bad_instantiate_missing_arg.output";
+           ocamlopt_byte_exit_status = "2";
            ocamlopt.byte;
 
-           flags = "-parameter Category -as-argument-for Category_b";
-           module = "category_b_of_category.mli category_b_of_category.ml";
+           compiler_reference = "bad_instantiate_missing_arg.reference";
+           check-ocamlopt.byte-output;
+         }{
+           module = "";
+           flags = "-instantiate";
+           program = "list_monoid-Float_list_element-Int_list_element.cmx";
+           all_modules = "list_monoid.cmx float_list_element.cmx int_list_element.cmx";
+           compiler_output = "bad_instantiate_repeated_param.output";
+           ocamlopt_byte_exit_status = "2";
+           ocamlopt.byte;
+
+           compiler_reference = "bad_instantiate_repeated_param.reference";
+           check-ocamlopt.byte-output;
+         }{
+           module = "";
+           flags = "-instantiate";
+           program = "not_the_correct_target_name.cmx";
+           all_modules = "list_monoid.cmx int_list_element.cmx";
+           compiler_output = "bad_instantiate_wrong_target_name.output";
+           ocamlopt_byte_exit_status = "2";
+           ocamlopt.byte;
+
+           compiler_reference = "bad_instantiate_wrong_target_name.reference";
+           check-ocamlopt.byte-output;
+         }{
+           module = "";
+           flags = "-instantiate -as-argument-for Monoid";
+           program = "list_monoid-Int_list_element.cmx";
+           all_modules = "list_monoid.cmx int_list_element.cmx";
            ocamlopt.byte;
 
-           flags = "-parameter Category -parameter Category_b -as-argument-for Category";
-           module = "product_category.mli product_category.ml";
+           module = "";
+           flags = "-instantiate -as-argument-for Monoid";
+           program = "monoid_of_semigroup-String_semigroup.cmx";
+           all_modules = "monoid_of_semigroup.cmx string_semigroup.cmx";
            ocamlopt.byte;
 
-           flags = "-parameter Semigroup -parameter List_element -w -misplaced-attribute";
-           module = "import_multi_arg.ml";
+           module = "";
+           flags = "-instantiate -as-argument-for Monoid";
+           program = "monoid_utils-Monoid_of_semigroup--String_semigroup.cmx";
+           all_modules = "monoid_utils.cmx monoid_of_semigroup-String_semigroup.cmx";
            ocamlopt.byte;
 
-           flags = "-parameter Semigroup -parameter List_element -w -misplaced-attribute";
-           module = "main_multi_arg.mli";
+           module = "";
+           flags = "-instantiate -as-argument-for Category";
```

Comment:
Note that `category_of_monoid.cmx` was already compiled with `-as-argument-for Category`. And here we say `-as-argument-for Category` again for the instance - do we check for some coherence?

Looking at `Instantiator.instantiate`, we simply take the `ui_arg_descr` of `Category_of_monoid.cmx`, and ignore the other one?

Reply:
Oh, good point. We probably don't need to say `-as-argument-for Category` here. (Possibly it should even be an error?)

Reply:
Yeah if we don't need it, we shouldn't allow it.

Reply:
Making a CR; doing it right would be more effort than we need for something we don't intend people to write yet

Reply:
(actually, I'll just do it but make it ugly for the moment)